gateway-sp-comms/src/shared_socket.rs

```diff
@@ -0,0 +1,712 @@
+                .lock()
+                .await
+                .contains_key(&peer_interface)
+            {
+                warn!(
+                    self.log, "discarding packet from unknown interface";
+                    "interface" => peer_interface.to_string(),
+                );
+                continue;
+            }
+
+            let data = &buf[..n];
+            let (header, kind, trailing_data) = match self.parse_message(data) {
+                Ok((header, kind, data)) => (header, kind, data),
+                Err(err) => {
+                    warn!(
+                        self.log, "failed to parse incoming packet";
+                        "data" => ?data,
+                        "peer" => %peer,
+                        "err" => %err,
+                    );
+                    continue;
+                }
+            };
+
+            let message = Message { header, kind };
+            if let Err(err) =
+                self.handle_message(&message, trailing_data, peer).await
+            {
+                warn!(
+                    self.log, "failed to handle incoming message";
+                    "message" => ?Message { header, kind },
+                    "peer" => %peer,
+                    "err" => %err,
+                );
+                continue;
+            }
+        }
+    }
+
+    fn parse_message<'a>(
+        &self,
+        data: &'a [u8],
+    ) -> Result<(Header, MessageKind, &'a [u8]), RecvError> {
+        // Peel off the header first to check the version.
+        let (header, remaining) = gateway_messages::deserialize::<Header>(data)
+            .map_err(RecvError::DeserializeHeader)?;
+        if header.version != version::V2 {
+            return Err(RecvError::VersionMismatch {
+                expected: version::V2,
+                sp: header.version,
+            });
+        }
+
+        // Parse the remainder.
+        let (kind, sp_trailing_data) =
+            gateway_messages::deserialize::<MessageKind>(remaining)
+                .map_err(RecvError::DeserializeBody)?;
+
+        Ok((header, kind, sp_trailing_data))
+    }
+
+    async fn handle_message(
+        &self,
+        message: &Message,
+        sp_trailing_data: &[u8],
+        peer: SocketAddrV6,
+    ) -> Result<(), RecvError> {
+        // Dispatch based on the kind of message. We handle host phase2 requests
+        // ourselves; any other SP request or response must be forwarded to a
+        // `SingleSp` instance (if one exists for `peer`).
+        match &message.kind {
+            MessageKind::MgsRequest(_) => {
+                Err(RecvError::InvalidMessageKind("MgsRequest"))
+            }
+            MessageKind::MgsResponse(_) => {
+                Err(RecvError::InvalidMessageKind("MgsResponse"))
+            }
+            MessageKind::SpRequest(SpRequest::HostPhase2Data {
+                hash,
+                offset,
+            }) => {
+                if !sp_trailing_data.is_empty() {
```

Comment:
Should we really ignore this, or is it better to drop the message as it's malformed. I feel like we shouldn't be liberal in what we accept here, but I'm open to argument.

Reply:
My only argument is a fairly weak one: we don't check for (or even log, which is inconsistent with the choice here!) empty trailing data on other messages that we don't expect any; e.g., https://github.com/oxidecomputer/management-gateway-service/blob/9e9fc718edea6756954447acddc9c47166f9ec41/gateway-sp-comms/src/single_sp.rs#L195, https://github.com/oxidecomputer/management-gateway-service/blob/9e9fc718edea6756954447acddc9c47166f9ec41/gateway-sp-comms/src/single_sp.rs#L218, and many more (search for `, _data`). A few options:

* Remove this check and log and silently accept unexpected trailing data like we do in all those others
* Add checks to all the others to log unexpected trailing data
* Change this to fail if we get unexpected trailing data and add similar check+fail to all the others

Seems like 3 is probably the way to go, but is big enough it should be a followup. What do you think?

Reply:
I think 3 is indeed the correct path. I also don't feel this is critical and it definitely can be done in a followup.

Reply:
Lifted this out to #49